### PR TITLE
Script for triggering build workflows

### LINKED ISSUE #15: Simplify (and document) triggering of rebuilds or updates
Eg. Windows FTBFS just was fixed https://git-annex.branchable.com/bugs/release_8.20201116_doesn__39__t_build_on_windows/ and I thought to trigger a rebuild (ideally just for windows).

I wonder what would be the most convenient way?  I am thinking about a script, which could also be used in workflows, e.g.
`tools/dispatch-workflows` or more specific `tools/dispatch-update` `tools/dispatch-build` which pretty much would be https://github.com/datalad/git-annex/pull/12/files#diff-90de5aca4b854158faaea471773df74b925ee576c00c8063220720b0a7e11985R37 
- default to that list of workflows to build across all OSes.
- token could be coded to come from `hub.oauthtoken` of `~/.gitconfig` so we could use locally as well
- should have a way to pass options pertinent to those workflows, e.g. to specify what committish to build

### PR DESCRIPTION
**Pull request by jwodder**
See the script's `--help` for usage information.

Closes #15.

### CODE REVIEW

File: .github/workflows/build-windows.yaml [hunk 1]
```diff
@@ -43,17 +43,14 @@ jobs:
...
+          if [ -n "${{ github.event.inputs.pr }}" ]
           then prnum="${{ github.event.inputs.pr }}"
                merge_sha="$(set -o pipefail; curl -fsSL -H "Authorization: bearer ${{ secrets.GITHUB_TOKEN }}" https://api.github.com/repos/datalad/git-annex/pulls/"$prnum" | jq -r .merge_commit_sha)"
                git fetch origin +"$merge_sha":refs/remotes/pull/"$prnum"/merge
                echo "BUILD_COMMIT=pull/$prnum/merge" >> "$GITHUB_ENV"
           elif [ -n "${{ github.event.inputs.commitish }}" ]
           then echo "BUILD_COMMIT=${{ github.event.inputs.commitish }}" >> "$GITHUB_ENV"
-          else echo No inputs provided to workflow dispatch
-               exit 2
+          else echo "BUILD_COMMIT=origin/upstream/master" >> "$GITHUB_ENV"
```
Comment (yarikoptic) [inline .github/workflows/build-windows.yaml:53]: note: all 3 workflows have identical code here?  we really should reduce code duplication, either via templates or just moving such code into helper scripts.

File: .github/workflows/tools/dispatch-build.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python3
```
Comment (yarikoptic) [inline .github/workflows/tools/dispatch-build.py:1]: let's strip `.py` suffix -- there is a shebang, and I do not see us invoking from windows, and windows people could just prepend with `python`. This is a typical recommendation in Debian land and, unless it is also to be used as python module, I tend to follow it

Comment (yarikoptic) [reply .github/workflows/tools/dispatch-build.py:1]: and shouldn't this script be used in `update-mirror.yml` (again - to not duplicate code; might need to get python  in the env if not there yet)

File: .github/workflows/tools/dispatch-build.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    def build_latest(self, ostypes=ALL_OS_TYPES, workflow_ref="master"):
+        for w in self.get_os_workflows(ostypes):
+            w.create_dispatch(ref=workflow_ref, inputs={})
+
+
+@click.command()
+@click.option(
+    "--os",
+    "ostypes",
+    type=click.Choice(["macos", "ubuntu", "windows"]),
```
Comment (yarikoptic) [inline .github/workflows/tools/dispatch-build.py:37]:
```suggestion
    type=click.Choice(ALL_OS_TYPES),
```
?

File: .github/workflows/build-windows.yaml (hunk 1 above)
Comment (jwodder) [reply .github/workflows/build-windows.yaml]: Helper script created.

**Comment by yarikoptic** [comment]
when dispatching those workflows, is there any way to provide "custom" additional information which would be visible in the dashboard?  e.g. I see that github is showing the branches for PRs in https://github.com/datalad/git-annex/actions?query=workflow%3A%22Build+git-annex+on+Ubuntu%22, may be somehow that could be passed into the dispatch event/call?

**Comment by yarikoptic** [comment]
asking just to figure out for starters which build corresponded to the one triggered for https://github.com/datalad/git-annex/pull/16 ... it feels that we cannot escape having some app/bot for those (unless somehow the dispatcher is also able to track across runs if any dispatched workflow is finished and report back to that PR)

File: .github/workflows/tools/dispatch-build.py (hunk 3 above)
Comment (jwodder) [reply .github/workflows/tools/dispatch-build.py]: Applied manually (because "Outdated suggestions cannot be applied.")

**Comment by jwodder** [comment]
@yarikoptic 
* There does not appear to be a way to show extra information on the dashboard.
* As I linked to [here](https://github.com/datalad/git-annex/issues/14#issuecomment-731296154), the runs triggered for #16 were: [Ubuntu](https://github.com/datalad/git-annex/actions/runs/374783240) (I fixed this failure), [macOS](https://github.com/datalad/git-annex/actions/runs/374783205), and [Windows](https://github.com/datalad/git-annex/actions/runs/374783216).

File: .github/workflows/tools/dispatch-build.py (hunk 2 above)
Comment (jwodder) [reply .github/workflows/tools/dispatch-build.py]: Done and done.

**Comment by yarikoptic** [comment]
An idea: since prs against annex aren't intended to be merged, we can introduce a change - instead of explicitly triggering workflow, we could push a commit with a copy of the workflow(s)

**Comment by jwodder** [comment]
@yarikoptic If we go with that, I'm assuming the workflow would push the build workflows once, when the PR is opened, and then it would know not to push to the PR again by ... checking whether `.github/workflows` exists on the PR's head branch?  Except (going just by Pygithub's API at the moment) I don't see a way to programmatically determine whether a given path exists in a branch.

**Comment by yarikoptic** [comment]
Thanks for checking!  Now it is weekend, woohoo, so let's digest - I think we might be doomed to just make a proper bot/app to make it non ugly

**Comment by yarikoptic** [comment]
comments addressed, duplication will be addressed separately.  Thank you @jwodder

**Pull request merged**